[proposal] Sharing code between Dev Container Features (#209)
devcontainers/spec: proposals/features-library.md

# Sharing code between Dev Container Features

ref: https://github.com/devcontainers/spec/issues/129

# Motivation

> The current model for authoring Features encourages significant copy and paste between individual Features.  Most of the Features published under the `devcontainers/features` namespace contain approximately 100 lines of [identifical helper functions](https://github.com/devcontainers/features/blob/main/src/azure-cli/install.sh#L29-L164) before the first line of the script is executed.  The process of maintaining these helper functions is tedious, error-prone, and leads to less readable Features.

Reviewer: ```suggestion
The current model for authoring Features encourages significant copy and paste between individual Features.  Most of the Features published under the `devcontainers/features` namespace contain approximately 100 lines of [identical helper functions](https://github.com/devcontainers/features/blob/main/src/azure-cli/install.sh#L29-L164) before the first line of the script is executed.  The process of maintaining these helper functions is tedious, error-prone, and leads to less readable Features.
```

Additionally, [members of the Feature authoring community have expressed interest in sharing code between Features](https://github.com/orgs/devcontainers/discussions/10).  Exploring the source code of self-published Features, it is evident that many authors have copy/pasted code from other Features.  This is a sign that there is a need for a more efficient way to share code between Features.

# Goal

Provide a generic and efficient way to share code between Features.

# Proposal

Add an `include` property to the `devcontainer-feature.json`.  This will be honored during the Feature [packaging stage](https://containers.dev/implementors/features-distribution/#packaging). The contents of the provided relative path is copied into the packaged Feature archive.

Property | Type | Description
--- | --- | ---

> `include` | `string[]` | An array of relative paths to package with the Feature. Paths are relative to root directory when packaging (often the git root directory).  If the element is a folder, it is copied recursively.

Reviewer 2: What would happen for paths that resolve outside of the feature root directory, such as `../shared/utils/`?

Author: Similar to how we don't resolve Local Features outside of the `.devcontainer` folder, I propose failing the packaging step.

Author: Added more detail

## Example

```json
{
    "name": "my-feature",
    "version": "0.0.1",
    "include": [
        "./utils/",
    ]
}
```

> The preceeding example will recursively copy the contents of the the `utils` folder into the temporary folder used to package the Feature.  The `utils` folder will be included with the published Features.

Reviewer: ```suggestion
The preceding example will recursively copy the contents of the `utils` folder into the temporary folder used to package the Feature.  The `utils` folder will be included with the published Features.
```

This functionality will enable sharing a common set of helper functions between Features.  The following example shows how a Feature could include a `utils` folder that contains a `common.sh` file.

```bash
# utils/common.sh
function common_function() {
   echo "Hello, '$1'"
}
```

```bash
# my-feature/install.sh

# Include common functions

> source "./utils/common.sh"

Reviewer 3: I guess this is `./utils` and not `../utils` because the packaged folder structure is different from the folder structure in the repository? So this can't be run from within the repository because the paths do not match?

Wouldn't it make development of features easier if the folder structures were the same? E.g., can we make it such that features can be run from the repository as "local" features during feature development?

@samruddhikhandale Any insights from your work on features?

Author: Yes, that's how I was originally imagining it.  Anything in the `includes` would be copied recursively to the root of the Feature's packaged directory.

Reviewer 3: I would suggest to keep the folder structure the same as in the repository. The main difficulty from what I see is the lookup of the `devcontainer-feature.json` which is currently expected at the root of the archive. If we know the feature id, we could additionally check in the folder with the same name as the unqualified id.

Author: For the Feature development case I don't think that makes too much of a difference - the `features test` command currently does a full copy into a temp folder inside of a temporary project's `.devcontainer` (to make it a 'local Features' from the perspective of the CLI).  This could likely be optimized, but we have strict rules over where local Features can live relative to the devcontainer.json, so it's a bit of hassle to _not_ do the copying.  I see the appeal of keeping the directory structure, and depending on if a 'local' Feature or a 'packaged' Feature, change where we look for the `include` files.  That said, we haven't really gotten feedback that performance on the testing command is an issue, and the vast majority of the time is not spent setting up the tests.

I don't know how useful that behavior is in practice, and makes figuring out the correct path slightly more confusing when trying to use an included script in a packaged Feature.

Also, how would it work sourcing from `../utils`.  Would we need to re-adjust how the entire archive is structured (ie: create a dummy `src` folder that has one folder `feature-name`, that contains `install.sh` and `devcontainer-feature.json`?

# Use common function
common_function "devcontainers"
```

-----

# Possible Extensions

This section provides hints to possible future work that could extend the proposal above.  These are not part of the proposal, but are provided to help the community understand the potential future impact of this proposal.

## Distributing shared library Features

In an effort to reduce the need for authors to copy/paste auxillary scripts between Feature namespaces, provide the ability to publish "library" Features.  These Features would be published [following the Feature distribution spec](https://containers.dev/implementors/features-distribution/#distribution) under a new manifest type. 

> A "library" can be included in a Feature's `include` property as an alterative to a relative path. By themselves, a "library" Feature would not be installable.

Reviewer 2: Similar, for this library feature, how would consumers locate these scripts on disk? Or would they be automatically sourced before execution of the consuming feature script?

Author: I think that's open for debate whether the user has the choice to source, or whether we try to auto-source whatever is included (might be trickier if we copy recursively, so perhaps we'd need to limit to include explicit script files)

Author: My example below has the consumer source explicitly at the top of their `install.sh`.

Author: I see that there are two options:

- No implicit sourcing.  The `include` simply copies the files into the archive.  It's up to the author to decide how and where to source/execute/copy those resources.
-  Treat script files (likely by looking for `.sh` and potentially other extensions) differently, and automatically source them in the feature-install.sh preamble script before executing the author's `install.sh`

Reviewer 3: No implicit sourcing is more flexible (e.g., scripts might be written like commands that immediately do something). 👍

> A "library" can be included in a Feature's `include` property as an alterative to a relative path. By themselves, a "library" Feature would not be installable.

Reviewer: ```suggestion
A "library" can be included in a Feature's `include` property as an alternative to a relative path. By itself, a "library" Feature would not be installable.
```

```

```bash
# src/featureA/install.sh

# Include common functions
source "../../utils/common.sh"

# Use common function
common_function "devcontainers"
```

## [Proposal B] Change: Structure of packaged Feature

> To simplify the proposed change above and keep the relative paths consistent before and after packaging, the structure of the packaged Feature will change to mirror the directory structure of the source code.

Author: Thanks, I removed some duplication.

When running the packaging step, the stage is provided a source directory for packaging >1 Feature, or the root of a project when packaging a single Feature.  The Feature contents will be placed into `./src/<FeatureID>`.  Other included directories will be copied into the root of the packaged artifact as indicated in **Proposal A**.

An implementation should resolve the `devcontainer-feature.json` by checking for its presence in the following order of precedence:

>  - `./src/<feature-id>/devcontainer-feature.json`

Reviewer 3: I wonder if we need to set `src` in stone. We could copy the `devcontainer-feature.json` to the root folder - probably updating relative paths in its properties, so they do not break. Or: Is it possible to have a symbolic link inside the tar file, making `devcontainer-features.json` at the root a link to the actual location?

Author: > I wonder if we need to set src in stone.

When publishing you're able to specify any folder as the 'src', ie:

`devcontainers features publish ~/josh/repos/features/src`

Is how i'd publish Features in a repo `features` with a folder `src` on my local machine.

_That_ repo could have any folder name instead of `src`.  I think, for simplicity in looking up the `devconatiner-feature.json`, we'd want to remap whatever that folder is to `src` (as I outlined above), but the user could keep that folder named to whatever they'd like it do be.

Not sure if that resolves your concern at all, just wanted to perhaps provide more context that the _user_ wouldn't necessarily need a folder named `src`, but internally we'd copy Feature's root folder (`<some-user-src-folder>/featureA/...` -> src/featureA...`

Author: > Is it possible to have a symbolic link inside the tar file, making devcontainer-features.json at the root a link to the actual location?

IMO that sounds like added complication.  Not sure, I'd need to do some digging to see how `tar` would react to this on different platforms.

>  - `./src/<feature-id>/devcontainer-feature.json`

Reviewer 3: I wonder if we need to set `src` in stone. We could copy the `devcontainer-feature.json` to the root folder - probably updating relative paths in its properties, so they do not break. Or: Is it possible to have a symbolic link inside the tar file, making `devcontainer-features.json` at the root a link to the actual location?
